frontend_tests/puppeteer_tests/03-compose.js

```diff
@@ -0,0 +1,168 @@
+const common = require('../puppeteer_lib/common');
+const assert = require("assert").strict;
+
+async function check_compose_form_empty(page) {
+    await common.check_form_contents(page, '#send_message_form', {
+        stream_message_recipient_stream: '',
+        stream_message_recipient_topic: '',
+        content: '',
+    });
+}
+
+async function close_compose_box(page) {
+    await page.keyboard.press('Escape');
+    await page.waitForSelector('#compose-textarea', {hidden: true});
+}
+
+async function compose_tests(page) {
+    await common.log_in(page);
+    await page.waitForSelector('#zhome .message_row', {visible: true});
+    const initial_msgs_count = await page.evaluate(() => {
+        return $('#zhome .message_row').length;
+    });
+
+    await common.send_multiple_messages(page, [
+        { stream: 'Verona',
+          topic: 'Reply test',
+          content: "We reply to this message",
+        },
+        { recipient: "cordelia@zulip.com",
+          content: "And reply to this private message",
+        },
+    ]);
+
+    assert.equal(await page.evaluate(() => {
+        return $('#zhome .message_row').length;
+    }), initial_msgs_count + 2);
+
+    await page.keyboard.press("KeyC");
+    await page.waitForSelector('#stream-message', {visible: true});
+    await check_compose_form_empty(page);
+
+    await close_compose_box(page);
+    await page.keyboard.press("KeyX");
+    await page.waitForSelector("#private_message_recipient", {visible: true});
+    await common.pm_recipient.expect(page, '');
+    await close_compose_box(page);
+
+    // Check that when you reply to a message it pre-populates the stream and topic fields
+    const message_we_reply_to = (await page.$x("//p[text()='We reply to this message']")).slice(-1)[0];
+    // we chose only the last element make sure we don't click on any duplicates.
+    await message_we_reply_to.click();
+    await common.check_form_contents(page,
+                                     '#send_message_form',
+                                     {
+                                         stream_message_recipient_stream: "Verona",
+                                         stream_message_recipient_topic: "Reply test",
+                                         content: "",
+                                     });
+    await close_compose_box(page);
+
+    // Check the same for a private message.
+    const priv_msg_we_reply_to = (await page.$x("//p[text()='And reply to this private message']")).slice(-1)[0];
```

Comment:
We also might want to extract a function that gets the last item of the array, instead of doing this `.slice(-1)[0]` everywhere. A better alternative would be to use `.pop()`  here but this actually removes the last item from the array so we may not be able to use it everywhere.